pkg/process/metadata/workloadmeta/workloadmeta.go

```diff
@@ -74,11 +85,30 @@ func (w *WorkloadMetaExtractor) Extract(procs map[int32]*procutil.Process) {
 	}
 
 	deadProcs := getDifference(w.cache, newCache)
-	w.grpcListener.writeEvents(deadProcs, newEntities)
 
 	w.cacheMutex.Lock()
 	w.cache = newCache
+	w.cacheVersion++
 	w.cacheMutex.Unlock()
+
+	// If previous cache diff hasn't been consumed, drop it
+	if len(w.diffChan) > 0 {
+		<-w.diffChan
+	}
```

Comment:
[race-condition] You're using a variation of [double checked locking pattern](https://en.wikipedia.org/wiki/Double-checked_locking) here which can cause this to block indefinitely. The channel has a chance of being consumed from between the size check and channel read. You should have a non-blocking channel read operation here instead avoid it

```
select {
case<-w.diffChan:
   // drop msg
default:
   // log and continue
}
```